nemo/collections/asr/data/data_simulation.py

```diff
@@ -979,6 +1000,164 @@ def _add_file(
+                                Note that using large value (greater than 3~4 sec) for `max_audio_read_sec` will slow down the generation process.
+                                If False, read the entire audio file.
+
+        Returns:
+            audio_file (torch.Tensor): Time-series audio data in a tensor.
+            sr (int): Sample rate of the audio file.
+            audio_manifest (dict): (modified) audio manifest dictionary.
+        """
+        audio_file_id = f"{audio_manifest['audio_filepath']}#{offset_index}"
+        if audio_file_id in buffer_dict:
+            audio_file, sr, audio_manifest = buffer_dict[audio_file_id]
+        else:
+            if read_subset:
+                audio_manifest = self._get_subset_of_audio_manifest(audio_manifest, offset_index)
+                segment = AudioSegment.from_file(
+                    audio_file=audio_manifest['audio_filepath'],
+                    offset=audio_manifest['offset'],
+                    duration=audio_manifest['duration'],
+                )
+            else:
+                segment = AudioSegment.from_file(audio_file=audio_manifest['audio_filepath'])
+            audio_file, sr = torch.from_numpy(segment.samples).to(self._device), segment.sample_rate
+            if read_subset and segment.duration < (audio_manifest['alignments'][-1] - audio_manifest['alignments'][0]):
+                audio_manifest['alignments'][-1] = min(segment.duration, audio_manifest['alignments'][-1])
+            if audio_file.ndim > 1:
+                audio_file = torch.mean(audio_file, 1, False).to(self._device)
+            buffer_dict[audio_file_id] = (audio_file, sr, audio_manifest)
+        return audio_file, sr, audio_manifest
+
+    def _get_random_offset_index(self, audio_manifest: dict) -> int:
+        """
+        Get an index for randomly accessing the silence in alignment timestamps. 
+
+        Args:
+            audio_manifest (dict): Audio manifest dictionary.
+                keys: 'audio_filepath', 'duration', 'alignments', 'words'
+
+        Returns: 
+            (int): Random offset index smaller than `offset_count`.
+        """
+        if len(audio_manifest['alignments']) <= self.min_alignment_count:
+            raise ValueError(
+                f"Audio file {audio_manifest['audio_filepath']} has less than {self.min_alignment_count} alignment timestamps."
+            )
+        # Find all silence indices
+        sil_inds = np.where((np.array(audio_manifest['words']) == '') == True)[0]
+        # If the audio file is shorter than the max_audio_read_sec, then we don't need to read a subset of the audio file.
+        if (
+            len(sil_inds) <= self.min_alignment_count
+            or (audio_manifest['alignments'][-1] - audio_manifest['alignments'][0]) < self.max_audio_read_sec
+        ):
+            return 0
+        else:
+            offset_index = np.random.randint(len(sil_inds) - 1)
```

Comment:
I wonder it this random selection will make the resulting audio very short? For example, the audio is 21s and `max_audio_read_sec==20`, the if `offset_index` is very close to the end of audio, the resulting sub-audio could be just 1s. Shall we change the random range so that the resulting audio is about `max_audio_read_sec` long? One thought is using binary search to find the longest random range:
```python
left = 0
right = len(sil_inds) - 1
while left < right:
    mid = left + (right - left) // 2
    dur_left = audio_manifest['alignments'][-1] - audio_manifest['alignments'][sil_inds[mid]]
    if  dur_left < self.max_audio_read_sec:
        right = mid - 1
    elif dur_left > self.max_audio_read_sec:
        left = mid + 1
    else:
        break
offset_index = np.random.randint(left + (right - left) // 2)
```
However, if it's ok to have very short sub-audio, we can just keep your current design.

Reply:
Reflected this binary search for securing `min_audio_read_sec`.
However, this still does not guarantee that we will always have enough words cause there are cases where the last word (alignment) is far away from the second last word. 
In this rare case, we need to use 1 word. Other cases are all handled.

+ Added reproducibility (seeding all the necessary randomizations) and checked now it is reproducible with the input seed.

+ Added a warning where 1 word is longer than `min_audio_read_sec`. This happens with forced alignment (libriSpeech or Fisher) where laughter and some long words are longer than 2.5 seconds.
We pop out a warning message then truncate the audio to have `min_audio_read_sec`.